File: finanze/domain/market_forecast.py
from typing import Any
from uuid import UUID

from domain.dezimal import Dezimal
from pydantic.dataclasses import dataclass


@dataclass
class MarketForecastPnlPoint:
    timestamp: int
    value: Dezimal
    currency: str
    entity_account_id: UUID | None = None
    wallet_address: str | None = None


@dataclass
class MarketForecastClosedPosition:
    currency: str
    entity_account_id: UUID | None = None
    wallet_address: str | None = None
    name: str | None = None
    market_key: str | None = None
    event_key: str | None = None
    outcome_key: str | None = None
    market_url: str | None = None
    icon_url: str | None = None
    outcome: str | None = None
    size: Dezimal | None = None
    avg_price: Dezimal | None = None
    price: Dezimal | None = None
    initial_value: Dezimal | None = None
    current_value: Dezimal | None = None
    cash_pnl: Dezimal | None = None
    percent_pnl: Dezimal | None = None
    cur_price: Dezimal | None = None
    redemption_value: Dezimal | None = None
    end_date: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    closed_at: str | None = None
    realized_pnl: Dezimal | None = None
    total_bought: Dezimal | None = None
    total_sold: Dezimal | None = None


@dataclass
class MarketForecastPnlAccountData:
    wallet_address: str
    currency: str
    profile: dict[str, Any] | None = None
    pnl_history: list[MarketForecastPnlPoint] | None = None


@dataclass
class MarketForecastClosedPositionsAccountData:
    wallet_address: str
    currency: str
    profile: dict[str, Any] | None = None
    closed_positions: list[MarketForecastClosedPosition] | None = None


@dataclass
class MarketForecastAccountSummary:
    entity_account_id: UUID
    entity_id: UUID
    account_name: str | None
    wallet_address: str
    currency: str
    profile: dict[str, Any] | None


@dataclass
class MarketForecastPnlAccount(MarketForecastAccountSummary):
    pnl_history: list[MarketForecastPnlPoint]


@dataclass
class MarketForecastClosedPositionsAccount(MarketForecastAccountSummary):
    closed_positions: list[MarketForecastClosedPosition]


@dataclass
class MarketForecastPnlResponse:
    interval: str
    accounts: list[MarketForecastPnlAccount]
    pnl_history: list[MarketForecastPnlPoint]


@dataclass
class MarketForecastClosedPositionsResponse:
    accounts: list[MarketForecastClosedPositionsAccount]
    closed_positions: list[MarketForecastClosedPosition]

File: frontend/app/src/types/marketForecast.ts
export type MarketForecastPnlInterval = "1d" | "1w" | "1m" | "all"

export type MarketForecastDecimal = number | string

export interface MarketForecastPnlPoint {
  timestamp: number
  value: MarketForecastDecimal
  entity_account_id?: string | null
  wallet_address?: string | null
}

export interface MarketForecastPosition {
  entity_account_id?: string | null
  wallet_address?: string | null
  title?: string | null
  slug?: string | null
  event_slug?: string | null
  icon?: string | null
  outcome?: string | null
  condition_id?: string | null
  asset?: string | null
  size?: MarketForecastDecimal | null
  avg_price?: MarketForecastDecimal | null
  price?: MarketForecastDecimal | null
  initial_value?: MarketForecastDecimal | null
  current_value?: MarketForecastDecimal | null
  cash_pnl?: MarketForecastDecimal | null
  percent_pnl?: MarketForecastDecimal | null
  cur_price?: MarketForecastDecimal | null
  redemption_value?: MarketForecastDecimal | null
  end_date?: string | null
  created_at?: string | null
  updated_at?: string | null
  closed_at?: string | null
  realized_pnl?: MarketForecastDecimal | null
  total_bought?: MarketForecastDecimal | null
  total_sold?: MarketForecastDecimal | null
}

export interface MarketForecastAccountSummary {
  entity_account_id: string
  entity_id?: string
  account_name?: string | null
  wallet_address: string
  profile?: Record<string, unknown> | null
}

export interface MarketForecastPnlAccount extends MarketForecastAccountSummary {
  pnl_history: MarketForecastPnlPoint[]
}

export interface MarketForecastClosedPositionsAccount extends MarketForecastAccountSummary {
  closed_positions: MarketForecastPosition[]
}

export interface MarketForecastPnlResponse {
  interval?: MarketForecastPnlInterval
  accounts: MarketForecastPnlAccount[]
  pnl_history: MarketForecastPnlPoint[]
}

export interface MarketForecastClosedPositionsResponse {
  accounts: MarketForecastClosedPositionsAccount[]
  closed_positions: MarketForecastPosition[]
}
